### PR TITLE
htlcswitch: ensure the packet queue exits on stop

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we ensure that the packet queue will always exit, by
continually signalling the main goroutine until it atomically sets a
bool that indicates its has been fully shutdown. It has been observed
that at times the main goroutine will wake up (due to the signal), but
then bypass the select and actually miss the quit signal, as a result
another signal is required. We'll continue to signals in a lazy loop
until the goroutine has fully exited.

### CODE REVIEW

File: htlcswitch/queue.go [hunk 1]
```diff
@@ -107,6 +117,7 @@ func (p *packetQueue) packetCoordinator() {
 			select {
 			case <-p.quit:
 				p.queueCond.L.Unlock()
+				atomic.StoreInt32(&p.streamShutdown, 1)
```
Comment (cfromknecht) [inline htlcswitch/queue.go]: Can this be defered?

File: htlcswitch/queue.go [hunk 2]
```diff
@@ -75,7 +78,14 @@ func (p *packetQueue) Start() {
...
 
-	p.queueCond.Signal()
+	// Now that we've closed the channel, we'll repeatedly signal the msg
+	// consumer until we've detected that it has exited.
+	for atomic.LoadInt32(&p.streamShutdown) == 0 {
+		p.queueCond.Signal()
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	p.wg.Wait()
```
Comment (cfromknecht) [inline htlcswitch/queue.go]: wait group no longer necessary

File: htlcswitch/queue.go (hunk 1 above)
Comment (Roasbeef) [reply htlcswitch/queue.go]: Yeah would be cleaner that way as well.

File: htlcswitch/queue.go [hunk 3]
```diff
@@ -137,10 +148,12 @@ func (p *packetQueue) packetCoordinator() {
 				atomic.AddInt64(&p.totalHtlcAmt, int64(-nextPkt.amount))
 				p.queueCond.L.Unlock()
 			case <-p.quit:
+				atomic.StoreInt32(&p.streamShutdown, 1)
 				return
 			}
 
 		case <-p.quit:
+			atomic.StoreInt32(&p.streamShutdown, 1)
```
Comment (halseth) [inline htlcswitch/queue.go]: would it make a difference to let the `quit` case be the first case in all the selects?

Comment (Roasbeef) [reply htlcswitch/queue.go]: Nope, the select case is selected randomly.

**Comment by Roasbeef** [comment]
PTAL.

File: htlcswitch/queue.go [hunk 4]
```diff
@@ -92,7 +100,7 @@ func (p *packetQueue) Stop() {
 // like reg congestion avoidance:
 //   * random dropping, RED, etc
 func (p *packetQueue) packetCoordinator() {
-	defer p.wg.Done()
+	defer atomic.StoreInt32(&p.streamShutdown, 1)
```
Comment (halseth) [inline htlcswitch/queue.go]: the `wg` variable can be removed now, I think?

**Pull request merged**